src/main/java/cuchaz/enigma/network/EnigmaServer.java

```diff
@@ -0,0 +1,240 @@
+package cuchaz.enigma.network;
+
+import cuchaz.enigma.gui.GuiController;
+import cuchaz.enigma.network.packet.KickS2CPacket;
+import cuchaz.enigma.network.packet.Packet;
+import cuchaz.enigma.network.packet.PacketRegistry;
+import cuchaz.enigma.network.packet.RemoveMappingS2CPacket;
+import cuchaz.enigma.network.packet.RenameS2CPacket;
+import cuchaz.enigma.translation.mapping.EntryMapping;
+import cuchaz.enigma.translation.mapping.EntryRemapper;
+import cuchaz.enigma.translation.representation.entry.Entry;
+
+import javax.swing.SwingUtilities;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class EnigmaServer {
+
+	// https://discordapp.com/channels/507304429255393322/566418023372816394/700292322918793347
+	public static final int DEFAULT_PORT = 34712;
+	public static final int PROTOCOL_VERSION = 0;
+	public static final String OWNER_USERNAME = "Owner";
+
+	private final int port;
+	private ServerSocket socket;
+	private List<Socket> clients = new ArrayList<>();
+	private Map<Socket, String> usernames = new HashMap<>();
+	private Set<Socket> unapprovedClients = new HashSet<>();
+
+	private final byte[] jarChecksum;
+
+	public static final int DUMMY_SYNC_ID = 0;
+	private final EntryRemapper mappings;
+	private Map<Entry<?>, Integer> syncIds = new HashMap<>();
+	private Map<Integer, Entry<?>> inverseSyncIds = new HashMap<>();
+	private Map<Integer, Set<Socket>> clientsNeedingConfirmation = new HashMap<>();
+	private int nextSyncId = DUMMY_SYNC_ID + 1;
+
+	private static int nextIoId = 0;
+
+	public EnigmaServer(byte[] jarChecksum, EntryRemapper mappings, int port) {
+		this.jarChecksum = jarChecksum;
+		this.mappings = mappings;
+		this.port = port;
+	}
+
+	public void start() throws IOException {
+		socket = new ServerSocket(port);
+		Thread thread = new Thread(() -> {
+			try {
+				while (!socket.isClosed()) {
+					acceptClient();
+				}
+			} catch (SocketException e) {
+				System.out.println("Server closed");
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		});
+		thread.setName("Server client listener");
+		thread.setDaemon(true);
+		thread.start();
+	}
+
+	private void acceptClient() throws IOException {
+		Socket client = socket.accept();
+		clients.add(client);
+		Thread thread = new Thread(() -> {
```

Comment:
Can we move such threads into a thread pool executor or so?

Reply:
As discussed on discord, I won't implement this change as these threads are long-running.